File: docker/rotorhazard/race_day.py
"""Mock **race-day autopilot** — emulate realistic races via the GridFPV test plugin.

Run this against a RotorHazard that has the ``gridfpv`` + ``gridfpv_mock`` plugins loaded (the
``cargo xtask race-day`` harness does this for you inside the dev container). It connects to RH's
socket.io, then **watches the race state**: whenever *you* stage + start a heat in the GridFPV
Director (driving RH into RACING), the autopilot emulates that heat's race over the wire —
injecting realistic per-pilot laps + RSSI bells through ``gridfpv_mock_pass`` on the nodes the
Director seated (those tuned to a channel). You drive the application; the test plugin drives the
races.

Pick a scenario (a "race-day personality") as argv[1]. Each shapes pace, spread, and the marshaling
edge cases (missed laps, false passes, a DNF) so you can practice different situations:

    clean    smooth, steady-pace laps, strong signal (the happy path)
    varied   per-pilot pace spread + lap-to-lap jitter (a realistic leaderboard)
    messy    marshaling practice: a missed lap, a false/extra pass, and a DNF
    pack     a tight field crossing close together (close finishes)

It loops indefinitely: every heat you run gets freshly emulated. Ctrl-C (or stopping the harness)
ends it. Re-running a heat re-emulates it.
"""

import random
import sys
import threading
import time

import socketio

RH_URL = "http://localhost:5000"

# RotorHazard race_status integers (see docker/rotorhazard/README.md): 1 = racing.
RACING = 1

# ---------------------------------------------------------------------------------------------
# Scenarios — each returns a per-node schedule of passes given the list of seated node indices.
# A schedule entry is (t_seconds_from_race_start, node_index, peak_rssi). The autopilot sorts the
# merged timeline and emits a gridfpv_mock_pass at each instant.
# ---------------------------------------------------------------------------------------------

BASELINE = 70
STRONG = 165
WEAK = 110
FALSE_PEAK = 100  # a spurious low bump a marshal should void


def _laps(node, count, lap_s, jitter, peak, start=0.6, rng=random):
    """A node's clean run: a holeshot then `count` laps at ~`lap_s` pace with ±jitter, all at
    `peak`. Returns [(t, node, peak), …] including the lap-0 holeshot."""
    out = []
    t = start
    for _ in range(count + 1):  # +1 for the holeshot (lap 0)
        out.append((t, node, peak + rng.randint(-6, 6)))
        t += lap_s * (1.0 + rng.uniform(-jitter, jitter))
    return out


def scenario_clean(nodes, rng):
    out = []
    for n in nodes:
        out += _laps(n, 4, 7.0, 0.06, STRONG, rng=rng)
    return out


def scenario_varied(nodes, rng):
    out = []
    for i, n in enumerate(nodes):
        pace = 6.0 + i * 0.9  # each pilot a bit slower than the last — a clear spread
        peak = STRONG - i * 8
        out += _laps(n, 5, pace, 0.18, max(peak, WEAK), rng=rng)
    return out


def scenario_messy(nodes, rng):
    """Marshaling practice: one node misses a lap, one gets a false/extra pass, the last DNFs."""
    out = []
    for i, n in enumerate(nodes):
        if i == len(nodes) - 1 and len(nodes) > 1:
            # DNF: only a holeshot + 2 laps, then nothing.
            out += _laps(n, 2, 7.5, 0.1, WEAK, rng=rng)
            continue
        laps = _laps(n, 4, 7.0, 0.1, STRONG, rng=rng)
        if i == 0:
            # Missed crossing: drop lap 2 (a gate miss the marshal must add back).
            laps = [p for k, p in enumerate(laps) if k != 2]
        if i == 1:
            # False pass: a low-peak bump ~1.5 s after the holeshot (a bounce/reflection to void).
            t0 = laps[0][0]
            laps.append((t0 + 1.5, n, FALSE_PEAK))
        out += laps
    return out


def scenario_pack(nodes, rng):
    out = []
    base_pace = 6.5
    for n in nodes:
        # Nearly identical schedules so the field crosses close together (dedup/ordering + close
        # finishes). Tiny jitter only.
        out += _laps(n, 4, base_pace, 0.03, STRONG, start=0.6, rng=rng)
    return out


SCENARIOS = {
    "clean": scenario_clean,
    "varied": scenario_varied,
    "messy": scenario_messy,
    "pack": scenario_pack,
}


# ---------------------------------------------------------------------------------------------
# The autopilot
# ---------------------------------------------------------------------------------------------


class RaceDay:
    def __init__(self, sio, scenario, seed=0):
        self.sio = sio
        self.scenario = scenario
        self.rng = random.Random(seed)
        self.worker = None
        self.stop = threading.Event()
        self.state_nodes = None
        self.state_evt = threading.Event()

    def tuned_nodes(self, timeout=3.0):
        """Ask the mock plugin for node state; return the indices tuned to a channel (the seats the
        Director set up for this heat)."""
        self.state_nodes = None
        self.state_evt.clear()
        self.sio.emit("gridfpv_mock_state")
        self.state_evt.wait(timeout)
        nodes = self.state_nodes or []
        return [n["index"] for n in nodes if n.get("frequency")]

    def on_state_ack(self, data):
        if data.get("action") == "state":
            self.state_nodes = data.get("nodes", [])
            self.state_evt.set()

    def start_race(self):
        self.stop_race()  # cancel any prior emulation
        self.stop.clear()
        self.worker = threading.Thread(target=self._run, daemon=True)
        self.worker.start()

    def stop_race(self):
        self.stop.set()
        w = self.worker
        if w and w.is_alive():
            w.join(timeout=1.0)
        self.worker = None

    def _run(self):
        nodes = self.tuned_nodes()
        if not nodes:
            print("race-day: no tuned/seated nodes found — did the heat seat any pilots?", flush=True)
            return
        for n in nodes:
            self.sio.emit("gridfpv_mock_reset", {"node": n})
        schedule = sorted(SCENARIOS[self.scenario](nodes, self.rng))
        print(
            f"race-day [{self.scenario}]: emulating {len(schedule)} passes across nodes {nodes}",
            flush=True,
        )
        t_start = time.monotonic()
        for t, node, peak in schedule:
            # Sleep until this pass's instant, bailing promptly if the race ended.
            while True:
                if self.stop.is_set():
                    print("race-day: race ended — stopping emulation", flush=True)
                    return
                ahead = t - (time.monotonic() - t_start)
                if ahead <= 0:
                    break
                time.sleep(min(ahead, 0.1))
            self.sio.emit(
                "gridfpv_mock_pass",
                {"node": node, "peak": int(peak), "baseline": BASELINE, "width": 14},
            )
        print("race-day: scenario complete — finish the heat in the Director when ready", flush=True)


def main():
    scenario = sys.argv[1] if len(sys.argv) > 1 else "clean"
    if scenario not in SCENARIOS:
        print(f"unknown scenario '{scenario}'. choose: {', '.join(SCENARIOS)}", flush=True)
        sys.exit(2)

    sio = socketio.Client(reconnection=True)
    autopilot = RaceDay(sio, scenario)

    @sio.on("gridfpv_mock_ack")
    def _ack(data):
        autopilot.on_state_ack(data)

    @sio.on("race_status")
    def _status(data):
        if data.get("race_status") == RACING:
            autopilot.start_race()
        else:
            autopilot.stop_race()

    sio.connect(RH_URL, wait_timeout=10)
    # No lap-minimum filter, so the brisk emulated laps all record.
    sio.emit("set_option", {"option": "MIN_LAP_TIME", "value": "0"})
    print(
        f"race-day autopilot ready [{scenario}] — stage + start a heat in the GridFPV Director and "
        "watch it race. Ctrl-C to stop.",
        flush=True,
    )
    try:
        sio.wait()
    except KeyboardInterrupt:
        autopilot.stop_race()
        sio.disconnect()


if __name__ == "__main__":
    main()

File: xtask/src/race_day.rs
//! `cargo xtask race-day` — the **mock race-day autopilot harness**.
//!
//! Stands up a RotorHazard with the GridFPV plugin **and** the test-only `gridfpv_mock` plugin, then
//! runs the [`race_day.py`](../../docker/rotorhazard/race_day.py) autopilot inside it. The autopilot
//! watches the race state and, whenever you stage + start a heat in the GridFPV Director, emulates
//! that heat's race over the wire (realistic per-pilot laps + RSSI bells via `gridfpv_mock_pass`).
//! You drive the application; the test plugin drives the races.
//!
//! - **`cargo xtask race-day [scenario] [--port P] [--container NAME]`** — ensure the race RH is up
//!   (build the image + `docker run` it if needed; no `mock_data` CSV — the autopilot is the only
//!   signal source), then exec the autopilot. Point the Director's active-event timer at the printed
//!   URL. Ctrl-C stops the autopilot (the RH container is left running for the next scenario).
//! - **`cargo xtask race-day list`** — the scenario menu.
//!
//! Scenarios (race-day "personalities"): `clean`, `varied`, `messy` (marshaling cases), `pack`.

use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::{Duration, Instant};

/// The persistent race-day RH container (distinct from the test harness's `gridfpv-rh-sig-*` and the
/// maintainer's `gridfpv-demo-rh`). Reused across runs; not auto-removed.
const RACE_CONTAINER: &str = "gridfpv-race-rh";
/// Default host port — matches the timer URL the deploy guide configures in the Director.
const DEFAULT_PORT: u16 = 5055;
/// The RH harness image (built from `docker/rotorhazard/`) — resolved through the testkit so
/// this can't drift from the tag the live harness builds. Race-day drives the Director by hand,
/// so it always uses the default RotorHazard version; the version × plugin sweep is
/// `cargo xtask live`'s job.
fn rh_image() -> String {
    gridfpv_testkit::rh_image_for(gridfpv_testkit::DEFAULT_RH_VERSION)
}

const SCENARIOS: &[(&str, &str)] = &[
    (
        "clean",
        "smooth, steady-pace laps, strong signal (the happy path)",
    ),
    (
        "varied",
        "per-pilot pace spread + lap-to-lap jitter (a realistic leaderboard)",
    ),
    (
        "messy",
        "marshaling practice: a missed lap, a false/extra pass, and a DNF",
    ),
    (
        "pack",
        "a tight field crossing close together (close finishes)",
    ),
];

pub fn run(args: &[String]) -> bool {
    let mut scenario = "clean".to_string();
    let mut port = DEFAULT_PORT;
    let mut container = RACE_CONTAINER.to_string();

    let mut it = args.iter();
    while let Some(arg) = it.next() {
        match arg.as_str() {
            "list" => {
                print_menu();
                return true;
            }
            "--port" => match it.next().and_then(|v| v.parse::<u16>().ok()) {
                Some(p) => port = p,
                None => {
                    eprintln!("--port needs a number");
                    return false;
                }
            },
            "--container" => match it.next() {
                Some(c) => container = c.clone(),
                None => {
                    eprintln!("--container needs a name");
                    return false;
                }
            },
            other if other.starts_with("--") => {
                eprintln!("unknown flag: {other}");
                return false;
            }
            name => scenario = name.to_string(),
        }
    }

    if !SCENARIOS.iter().any(|(n, _)| *n == scenario) {
        eprintln!("unknown scenario: {scenario}");
        print_menu();
        return false;
    }
    if !docker_present() {
        eprintln!("\x1b[31mdocker not found on PATH.\x1b[0m race-day drives a real container.");
        return false;
    }

    let root = workspace_root();
    if !ensure_image(&root) {
        return false;
    }
    if !ensure_container(&root, &container, port) {
        return false;
    }

    // Copy the autopilot in fresh (so edits land without rebuilding the container).
    let script = root.join("docker/rotorhazard/race_day.py");
    if !run_ok(
        "docker",
        &[
            "cp",
            script.to_str().unwrap(),
            &format!("{container}:/tmp/race_day.py"),
        ],
    ) {
        eprintln!("failed to copy race_day.py into {container}");
        return false;
    }

    println!(
        "\n\x1b[1mMock race day\x1b[0m — scenario \x1b[1m{scenario}\x1b[0m on RotorHazard \
         \x1b[1mhttp://localhost:{port}\x1b[0m (container `{container}`).\n"
    );
    println!("\x1b[1mNext steps:\x1b[0m");
    println!(
        "  1. In the GridFPV Director, point your active event's RotorHazard timer at \
         \x1b[1mhttp://localhost:{port}\x1b[0m and make the event active (timer reads Connected)."
    );
    println!(
        "  2. Build an event with pilots + heats, seat pilots on nodes, then \x1b[1mStage → Start\x1b[0m a heat."
    );
    println!(
        "  3. Watch it race — the autopilot below emulates the seated nodes per the scenario."
    );
    println!(
        "  4. Marshal / advance as you like; run the next heat. Ctrl-C here ends the autopilot.\n"
    );

    // Exec the autopilot in the foreground (streams its log; runs until Ctrl-C).
    run_ok(
        "docker",
        &["exec", &container, "python3", "/tmp/race_day.py", &scenario],
    )
}

fn print_menu() {
    println!("\n\x1b[1mMock race-day scenarios\x1b[0m (cargo xtask race-day <name>):\n");
    for (name, blurb) in SCENARIOS {
        println!("  \x1b[1m{name:<8}\x1b[0m {blurb}");
    }
    println!("\nEach watches the race state and emulates every heat you Start in the Director.");
}

/// Build the RH harness image if it isn't present locally — the testkit's own auto-build, so
/// there is exactly one place that knows how to build it (tag, build-arg, and context alike).
fn ensure_image(_root: &Path) -> bool {
    gridfpv_testkit::ensure_rh_image(gridfpv_testkit::DEFAULT_RH_VERSION);
    true
}

/// Ensure the race RH container is running with BOTH plugins mounted (and no `mock_data` CSV, so the
/// autopilot is the only signal source). Reuses an already-running container; starts one otherwise.
fn ensure_container(root: &Path, container: &str, port: u16) -> bool {
    let running = Command::new("docker")
        .args(["inspect", "-f", "{{.State.Running}}", container])
        .output()
        .map(|o| o.status.success() && String::from_utf8_lossy(&o.stdout).trim() == "true")
        .unwrap_or(false);
    if running {
        println!("Reusing running RotorHazard container `{container}`.");
        return true;
    }
    // Remove any stopped leftover of the same name, then start fresh.
    let _ = Command::new("docker")
        .args(["rm", "-f", container])
        .output();

    let gridfpv = root.join("plugins/gridfpv");
    let mock = root.join("plugins/gridfpv_mock");
    let mount = |dir: &Path, name: &str| {
        format!(
            "{}:/opt/RotorHazard/src/server/plugins/{name}:ro",
            dir.display()
        )
    };
    println!("Starting RotorHazard `{container}` on :{port} (gridfpv + gridfpv_mock plugins)…");
    let ok = run_ok(
        "docker",
        &[
            "run",
            "-d",
            "--name",
            container,
            "-p",
            &format!("{port}:5000"),
            "-v",
            &mount(&gridfpv, "gridfpv"),
            "-v",
            &mount(&mock, "gridfpv_mock"),
            &rh_image(),
        ],
    );
    if !ok {
        eprintln!("failed to start {container}");
        return false;
    }
    // Wait for the HTTP port to accept connections.
    let deadline = Instant::now() + Duration::from_secs(60);
    while Instant::now() < deadline {
        if std::net::TcpStream::connect(("127.0.0.1", port)).is_ok() {
            std::thread::sleep(Duration::from_secs(3)); // settle the socket API
            return true;
        }
        std::thread::sleep(Duration::from_millis(500));
    }
    eprintln!("RotorHazard did not open port {port} in time");
    false
}

fn workspace_root() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("xtask manifest dir has a parent (workspace root)")
        .to_path_buf()
}

fn docker_present() -> bool {
    Command::new("docker")
        .arg("--version")
        .output()
        .map(|o| o.status.success())
        .unwrap_or(false)
}

fn run_ok(program: &str, args: &[&str]) -> bool {
    Command::new(program)
        .args(args)
        .status()
        .map(|s| s.success())
        .unwrap_or(false)
}
